packages/mcfunction/src/parser/macro.ts

```diff
@@ -0,0 +1,153 @@
+import * as core from '@spyglassmc/core'
+import { localeQuote, localize } from '@spyglassmc/locales'
+import type { MacroChildNode, MacroKeyNode, MacroNode } from '../node/index.js'
+
+/**
+ * Parse a macro.
+ */
+export function macro(): core.Parser<MacroNode> {
+	return (
+		src: core.Source,
+		ctx: core.ParserContext,
+	): core.Result<MacroNode> => {
+		const start = src.cursor
+		const children: MacroChildNode[] = []
+
+		// Handle the starting '$'
+		if (src.peek() == '$' && src.peek(2) !== '$(') {
+			const ans: MacroChildNode = {
+				type: 'mcfunction:macro_child',
+				range: core.Range.create(start, start + 1),
+				options: { type: 'sign' },
+				value: '$',
+			}
+			src.skip()
+			children.push(ans)
+		}
+
+		// Handle the rest of the line
+
+		let beginning = src.cursor
+		let txt = src.readUntil('$', '\r', '\n')
+		do { // Do-while because we always want it to work the first time
+			let wasMacro = false
+			const child: MacroChildNode = parse(src, ctx, txt, beginning, wasMacro)
+			if (child.options.type === 'macro') {
+				wasMacro = true
+			}
+			// Handle a non-macro $ in the middle of the line
+			let subMacro: MacroChildNode | undefined
```

Comment:
Why do we need to handle non-`$` in the middle of the line? Those could just be part of the command. What are sub macros?

Reply:
In a command like `$scoreboard players set $mode settings $(value)` the `$mode` would mess with the existing parsing, a sub macro is a macro that gets caught in the rest of the command as a by-product of that problem and is extracted with the following code

Reply:
What if you have the line `$tellraw $(players) ["$ Hello everyone! $"]`?

Reply:
![image](https://github.com/SpyglassMC/Spyglass/assets/38361803/9d1b84d7-e322-4655-b033-67977ce86fe9)
Currently works okay

Reply:
Yeah I get that it "works" okay, but I'm more concerned in how it works. (I edited my previous comment afterwards)

Reply:
![image](https://github.com/SpyglassMC/Spyglass/assets/38361803/c9cbd06e-9192-441e-a734-2f81995314fb)
This is what the subtree looks like